### ISSUE TITLE
when can a different type be passed to an out formal?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#16582]
Spin-off from #16554.

Should candidate selection in function resolution require that an actual argument for an `out` formal have a matching type?

For example, consider this program:

``` chapel
proc fInt8Out(out arg: int(8)) { }
var myInt: int = 1;
fInt8Out(myInt); // should this resolve?
```

Consider also this program with a custom record:

``` chapel
record R {
  var x: int = 0;
  proc init=(x: int) {
    writeln("In init=");
    this.x = x;
  }
}

proc =(ref r: R, x: int) {
  writeln("In assign");
  r.x = x;
}


proc foo_out(out x: int) {
  x = 45;
}

proc foo_ret(): int {
  return 33;
}

{
  var a:R;
  a = foo_ret(); // works today - split initializes and calls R.init=
  var b:R = new R();
  b = foo_ret(); // works today - calls = overload
}

{
  var c:R;
  foo_out(c); // Doesn't resolve today due to arg mismatch
  var d:R = new R();
  foo_out(d); // Doesn't resolve today due to arg mismatch
}
```

The question is, should the *c* and *d* cases be allowed?

I know of 4 general design directions here:

1. Do not allow calls to resolve for an `out` formal when the actual type does not match (where matching can include generic instantiation)
2. Allow all actual arguments to resolve for an `out` formal but then produce a compilation error for an unresolved `init=` or `=` between the types if one is not present
3. During candidate selection, check for the availability of an `init=` and/or `=` between the actual type and the `out` formal type.
4. Allow the call to resolve if the `out` formal's type can implicitly convert to the actual type (e.g. with #16729).

**Comment by mppf** [comment]
Supposing we are doing Option 3, do we check for `init=`, `=`, or both? For same-type `init=` and `=` we opted to make it an error if both are not provided (see PR #14956 and discussion in #8065). In issue #15838 we are considering extending that checking to also cover different-type cases or potentially generating `init=` from `=` or vice versa.

What is the impact on this issue? All four of the proposed directions in #15838 would lead us to only needing to check for one of `init=` or `=` when computing if a function using `out` intent with a different type is a candidate.

**Comment by mppf** [comment]
See https://github.com/chapel-lang/chapel/issues/15838#issuecomment-708463962 - depending on how we handle non-record types, a cross-type `=` might cause us to allow implicit conversions for an `out` intent argument.

**Comment by mppf** [comment]
I edited the issue's description to add this example:

``` chapel
proc fInt8Out(out arg: int(8)) { }
var myInt: int = 1;
fInt8Out(myInt);
```

This does not compile today but if we decide to allow automatic conversions for `out` then it should work, because `int(8)` can convert to `int`.

**Comment by bradcray** [comment]
> Supposing we are doing Option 3, do we check for `init=`, `=`, or both?

I was imagining that if the actual argument is uninitialized (a candidate for split-init), we'd check for `init=` and if it had been initialized (was not a candidate for split init) we'd check for `=`.

**Comment by mppf** [comment]
We have been developing a viewpoint that implicit conversions should be something one can request separately from `=` / `init=`. At the same time, it seems that implicit conversions are something one should opt in to (rather than opt out of). As a result, existence of`=` / `init=` shouldn't impact candidate selection and therefore we can rule out Option 3.

**Comment by mppf** [comment]
I have added Option 4 -- Allow the call to resolve if the out formal's type can implicitly convert to the actual type (e.g. with #16729).

**Comment by mppf** [comment]
One interesting implication of Option 2 is that it would enable us to support type-less split-init for `out`, e.g.

``` chapel
var x, y, z;
var res = foo(x,y,z)

proc foo(out x:Type1, out y:Type2, out z:Type3): Type4 { ... }
```

It would make the type flow more like return - the type of these variables is provided by the function `foo`.

I wonder if this approach will run into problems with type inference. With the return type, we put a fair amount of effort in to making return types be inferred so they don't have to be specified in detail. With the `out` intent, inference can also occur, but it comes from the call site. If we made this change, the function `foo` would perhaps need to infer the types of x/y/z from the function body (just as it would a return) in the event that their types are not declared?

**Comment by mppf** [comment]
> however, we've definitely had users express interest in having the callsite play more of a role in choosing between overloads (though I worry about the instability that it could lead to...)

Just recording this thought somewhere... Going all the way towards choosing a different overload based the type it returns (say, one returning a tuple and one returning an int) seems really problematic for type inference. `var x = f()` could have `x` be a tuple or not a tuple. Even  if we said that was an ambiguity error, then to call `f` you'd have to always write the return type like `var x: int = f()` or `var x: (int, int) = f()` which seems like it's making productivity worse rather than better.

**Comment by bradcray** [comment]
> One interesting implication of Option 2 is that it would enable us to support type-less split-init for out

This is pretty compelling to me.  There was a case this week when I was thinking about changing a routine of the form:

```chapel
var (x, y, z) = foo();
```

to:

```chapel
var res = foo(x, y, z);
```

but then I found myself thinking "Oh, but then I'd have to define the types of `x`, `y`, and `z`, which seems annoying."  But if I didn't need to and could just type:

```chapel
var x, y, z;
var res = foo(x, y, z);
```

then that seems nice and productive, and let's the choice between returning or "passing by out" to be a style choice rather than resulting in a different in capabilities (which builds on the theme we've been building towards lately of having them be more similar).

[Of course, the other way to make them more similar is to have the return type play a role in candidate selection, as out arguments currently do, but that gives me concerns like the ones you list above.  A case I particularly worry about is `bar(foo())` where both bar and foo have multiple overloads such that either one could influence which one was called.  Or `bar(foo(), baz())` where the negotiation between candidates gets even more complex.  This is what I meant by "the instability it could lead to" above.]

> I wonder if this approach will run into problems with type inference.
...
> If we made this change, the function foo would perhaps need to infer the types of x/y/z from the function body (just as it would a return) in the event that their types are not declared?

The "just as it would a return" seems reassuring to me (which isn't to say there's not work involved, but...).

If we think of:

```chapel
proc foo(out x: int) { x = expr; }
```

as being similar to:

```chapel
proc foo(): int {
  var retval: int;
  retval = expr;
  return retval;
}
```

or:

```chapel
proc foo(out x) { x = expr; }
```

as being similar to:

```chapel
proc foo() {
  var retval;
  retval = expr;
  return retval;
}
```

then they seem fairly symmetric?  (I'm trying to remember how we currently unify the types of multiple return statements, but however that works, it seems that it could similarly apply if there were multiple assignments to a type-inferred `out` argument?)

**Comment by mppf** [comment]
> then they seem fairly symmetric?

Yeah

> (I'm trying to remember how we currently unify the types of multiple return statements, but however that works, it seems that it could similarly apply if there were multiple assignments to a type-inferred out argument?)

It happens in `resolveReturnTypeAndYieldedType`where we basically examine all `return` statements.

An `out` argument has the potential to not be set. Today that leads to default initialization but with a typeless `out` argument it would be an error:

``` chapel
proc f(out x) {
  return; // uh-oh, can't establish the type of x
}
```

But presumably we could establish the type by setting it from a call expression using return or out:

``` chapel
proc g(out x, out y) {
  x = generateValue();
  returnValueOut(y);
  return; // OK
}
```

One difference `out` has with `return` is that it could be a separate statement setting the `out` value from the one returning. As a result this little inference analysis would need to be more flow sensitive:

``` chapel
proc h(out x) {
  if something {
    x = generateValue();
  } else {
    x = generateValueAnotherWay();
  }
  return;
}
```

Here I think existing rules already cover it if we imagine these assignments to `x` as candidates for split init or for default-init and assignment. That is, the compiler would think of it as this:

``` chapel
proc h(out xFormal) {
  var x;
  if something {
    x = generateValue(); // ok - split init
  } else {
    x = generateValueAnotherWay(); // ok - split init
  }
  move xFormal, x
  return;
}
```

None of that makes me want to back away from the approach. I think this approach is interesting. It will just require some non-trivial implementation adjustment to try.

**Comment by bradcray** [comment]
> with a typeless out argument it would be an error:

I agree with this.  A typeless out argument would need to be set.

> As a result this little inference analysis would need to be more flow sensitive:

Your specific example seems as though it's equivalent to having a `return` on each of the two branches, isn't it?  That said, I could definitely believe that there are cases which are not equivalent to having a `return` on multiple control paths.

Having noted that, it's interesting to realize that in the presence of conditionals, our unification of types for return statements is pretty different than how we handle split initialization in conditionals.  E.g., compare:

```chapel
proc foo(x: bool) {
  if x then
    return 42;
  else
    return 4.2;
}

writeln(foo(true));
writeln(foo(false));
```

to:

```chapel
proc foo(x: bool) {
  var y;

  if x then {
    y = 42;
  } else {
    y = 4.2;
  }
  return y;
}

writeln(foo(true));
writeln(foo(false));
```

**Comment by mppf** [comment]
> Having noted that, it's interesting to realize that in the presence of conditionals, our unification of types for return statements is pretty different than how we handle split initialization in conditionals. E.g., compare:

Split init is more restrictive and AFAIK we could relax it in the future if we wish to (as a non-breaking change).

I'm trying to understand further if Option 2 - _Allow all actual arguments to resolve for an out formal but then produce a compilation error for an unresolved `init=` or `=` between the types if one is not present_ - would have any surprising impact on existing patterns using the `out` intent in modules and tests.

I used `git grep '[(,] *out  *'` to look at uses of `out` intent. I particularly focused on code that is not testing something about `out` intent itself.

Most uses of `out` intent already use a concrete type and I haven't yet noticed a case where overload resolution is important on the `out` argument itself.

Here are a few interesting cases:

* `proc InitializeEdges(nEdges: int, out nNodes: int, U, V, W)` - note here that `nNodes` is concrete even though other arguments are generic. Maybe that's because its type was easy to express. It's very common for the type of an `out` argument to be a built-in type (e.g. int, string, etc).
* `proc channel.readbits(out v:integral, nbits:integral):bool throws` - this is one of the few cases where the `out` argument is meaningfully generic. If we were to make it concrete, by making overloads for all the integral types like `int(?w)` and `uint(?w)`, then this function would rely on overload resolution in a way that is incompatible with Option 2. But Option 2 should work fine with it as written.
* `proc rangecast(out r1: range(?), r2: range(?)): void` - this one is a generic `out` argument with a type restriction but the type restriction isn't particularly important - the function has a pretty unique name and the type of the `out` argument need not be considered when resolving it. An interesting note about it is that we currently write cast functions as `_cast(type t: toType, from: fromType)` but we could conceivably write them as `_cast(out dstVariable: toType, from: fromType)`. If we wanted to support such a signature for cast-like functions we wouldn't be able to do Option 2.

Anyway I haven't found anything that makes me think we need to rule out Option 2.

I think that the next step is to try adjusting function resolution to ignore the type of `out` arguments when doing disambiguation. This could be done (as an experiment) without impacting the current flow of types for `out` formals from the call site to the called function. From there the next step would be to adjust the way that the type for generic `out` formals is inferred (to be inferred from the called function rather than from the call site).

**Comment by mppf** [comment]
grepping similarly in Arkouda I found also this case:
`inline proc copyElement(out dst: ?t, in src: t) where isNumeric(t)`

This case is interesting because if the type of the `out` argument is inferred from the function body, that means that its type is not known before the type of the `in src` argument is. That causes a problem with the ordering of `?t` vs `t`. If it were

`inline proc copyElement(in src: ?t, out dst: t) where isNumeric(t)`

then there would be no issue (the `out dst: t` would be a kind of type assertion on the body of the function).

Given that we are expecting the `out` argument to do a move/`=` at the call site, I think it's reasonable for a function like this to just leave out the type restriction on the `dst` argument; e.g. `inline proc copyElement(out dst, in src: ?t) where isNumeric(t)`.

**Comment by mppf** [comment]
I suppose one could argue that, under Option 2, an `out` argument using a cross-type `=` at the call site is a kind of implicit conversion. For example:

``` chapel
record MyRecord { }
proc = (ref lhs: MyRecord, const ref rhs: string) { ... }

proc returnsStringThroughOut( out s: string ) { s = "hello"; }

var x: MyRecord = new MyRecord();
returnsStringThroughOut(x); // (1)
```

at (1), the call resolves to the function above (because the type of the `out` argument is ignored in choosing candidates) and then the `out` argument is handled at the call site with `=` in this case. The `=` assigns from a string to a record. Is this an "implicit conversion"? 

It seems to me that if a user not wanting to use implicit conversions is happy to write

``` chapel
var y: MyRecord = "hello";
```

that it is similarly OK for `returnsStringThroughOut(x)` above to use the `= ( MyRecord, string )`.

It also seems to me that this is not "risky" like the other cases of implicit conversion for calls (such as with `in` intent) because there can be no surprise about which function is called. The `out` formal does not participate in that choice but that fact has nothing to do with the availability of conversions.

So, I think Option 2 is compatible with the idea of allowing users to avoid confusion about which function is called by avoiding user-defined implicit conversions.

**Comment by mppf** [comment]
> I think that the next step is to try adjusting function resolution to ignore the type of out arguments when doing disambiguation. This could be done (as an experiment) without impacting the current flow of types for out formals from the call site to the called function.


I did this and saw only two failures in tests that have to do with testing `out` intent behavior
```
[Error matching compiler output for classes/vass/ref-like-intents/out-subclass]
[Error matching compiler output for classes/vass/ref-like-intents/out-superclass]
```

This seems to indicate we don't have code that was relying on disambiguation of `out` arguments based on type.

**Comment by mppf** [comment]
> * `proc channel.readbits(out v:integral, nbits:integral):bool throws` - this is one of the few cases where the `out` argument is meaningfully generic. If we were to make it concrete, by making overloads for all the integral types like `int(?w)` and `uint(?w)`, then this function would rely on overload resolution in a way that is incompatible with Option 2. But Option 2 should work fine with it as written.

Actually, I was wrong about this case. If the type information flows from the called function, as with `return`, then `channel.readbits(myInt8, 4)` will no longer work as-is. We could add a type argument or have the type of`nbits` imply the type of the result.

I think that the easy way to fix the function is to use `ref v: integral` rather than `out v: integral`.

**Comment by mppf** [comment]
This change interacts with the way that array formal types are generic:

``` chapel
proc foo(out B: [1..3] int) {
  for i in B.domain do
    B(i) = i;
}
```

Supposing we have the type information flow from the function (like with returns). If the array type is generic,  then this `out` argument is never initialized enough to give it a type. If it is not generic, then this `out` argument cannot be used to initialize (say) a Block distributed array. (The code would have to use `ref` instead). Like the `readbits` example I think this gets to the conceptual question - is `out` closer to a decorator meant to document what is going on with a `ref` argument that is always set - or is it something fundamentally different from `ref` ?

I could adjust this `out` formal to not consider the type to be generic (since it is more similar to a variable declaration) or we could error on this pattern.

**Comment by mppf** [comment]
> If it is not generic, then this out argument cannot be used to initialize (say) a Block distributed array.

That's sortof wrong; a Block argument can be initialized with `=` at the call site to implement the `out` argument.

**Comment by mppf** [comment]
Here is another odd case:
``` chapel
proc test_out(out xxx...)
```

we could have the number of tuple elements be determined by the caller but their types be determined by the called function but that seems a bit too strange to do in my opinion, so I'd just prohibit `out` varargs functions.

**Comment by mppf** [comment]
Another odd case is code like this

``` chapel
proc f(out arg) {
  if something
    arg = 1;
}
```

This came up in some SSCA2 code. Here the types are left out for convenience rather than for any good reason. The split-init rules don't let arg be default inited in this case, so the code would have to change (`ref arg`, add a type to it, or initialize it also in an else block).

**Comment by mppf** [comment]
The draft PR #16990 demonstrates that the Option 2 and types being inferred from the called function is a reasonable enough direction. Some tests did need to change. At this point I think we have the evidence from the implementation to make a decision and I will ask for more dicussion.

**Comment by bradcray** [comment]
> It also seems to me that this is not "risky" like the other cases of implicit conversion for calls (such as with in intent) because there can be no surprise about which function is called. The out formal does not participate in that choice but that fact has nothing to do with the availability of conversions.

This resonates strongly for me.

> I think that the easy way to fix the function is to use ref v: integral rather than out v: integral.

Making readbits() take its argument by `ref` makes sense to me.  Specifically, its purpose feels more like "take this pre-existing thing and populate it for me" than "create something from nothing".  And if someone really wanted the latter (say to initialize a variable), then I think it's reasonable to have them create the version that takes an explicit `type` as an argument (or can infer it from some other context) is reasonable.  Finally, since we're just talking about singleton `int`s here, I don't think it's a big deal if the result is to default-init the int and then pass it in (whereas if we were talking about a huge data structure, that might be more problematic).

For this case:

```chapel
proc foo(out B: [1..3] int) {
```

I'd be inclined to interpret `[1..3] int` more like `var A: [1..3] int;` (that is, we're literally creating a default array) and less like formal argument `A: [1..3] int, ...` (which is to say, generic).  That means if I wanted the formal to be block distributed, I think it should be declared in a way that makes it apparent it is; or equivalently, that I think the normal "DR formal means generic w.r.t. domain map" rule wouldn't apply for `out` arguments.  And I agree with your point that the DR can be assigned back to a Block formal of same size without problem.  Maybe put a different way:  If the `out` formal were so large that it would be ridiculous for it to be a DR array, then I think it's also reasonable to declare it in such a way that it's distributed nature were explicit rather than inferred from the callsite in some way.

> so I'd just prohibit out varargs functions.

That proposal doesn't give me any pause.

For this case:

```chapel
 proc f(out arg) {
  if something
    arg = 1;
}
```

am I understanding correctly that if there were an else clause that assigned/initialized `arg` it'd be OK?  Or if `arg` were assigned before the conditional?  (but probably not if it were initialized after the conditional?).  I'm fine with that and think it makes sense given the other rules we have.  To me, this is pretty similar to having a single `return 1;` within that conditional with no other returns.

Overall, I feel good about option 2 based on this analysis, and because it makes `out` more equivalent to `return`.  (In fact, let me put the question to you:  If someone wanted to understand the differences between `out` and `return`, what would they be?  One case that occurs to me is that if I was correct about the previous example, it would be legal to put a `return` after the conditional.  That makes me wonder if there's still something pretty different about how return types are inferred vs. out arguments).

**Comment by mppf** [comment]
Re ` proc f(out arg) {` --

> am I understanding correctly that if there were an else clause that assigned/initialized arg it'd be OK? Or if arg were assigned before the conditional? (but probably not if it were initialized after the conditional?).

Yes, according to the normal split init rules.

>  If someone wanted to understand the differences between out and return, what would they be?

* `out` is syntactically declared and assigned to vs `return` is a keyword where returning happens
  * as a result, lack of a `return` on some control flow path means nothing is returned - and is an error if other paths do return - but not setting a typed `out` formal merely causes it to be default initialized. In other words, an `out` formal with a type can be read from - but such a thing is not possible with the `return` value (at least not within the function doing the returning).
* `out` arguments are named and you can have many of them vs `return` which is unnamed and there can only be one returned value

(We would also need to say something about the ways that split-init type inference is more restrictive than the current return type inference. I didn't emphasize that here because split-init could become more flexible over time so am not so sure it makes a big difference for the current decision.).

> That makes me wonder if there's still something pretty different about how return types are inferred vs. out arguments

I think there are fundamental differences (described above) and also not fundamental differences (like type inference behavior of split-init vs return type inference).

**Comment by vasslitvinov** [comment]
Since we find this pattern attractive:
```chpl
var myVar: MyType;              // not initialized yet
initializeViaOutIntent(myVar);  // split-initialization here
```

it would be unfortunate if we disallowed the possibility for the out-formal of initializeViaOutIntent() to be generic and get its concrete type from the actual. The user would have to write ex.
```chpl
initializeViaOutIntent(myVar, myVar.type);
```

To support the elegant version, I propose to provide syntax for inferring the concrete type of an out-formal from the actual argument. For example, using a query type:
```chpl
proc initializeViaOutIntent(arg: ?T) {   // arg.type is determined by the actual
  .....
}
```

Such a formal would still not participate in candidate selection, as per the leading proposal.

This would enable the scenarios mentioned above:
* channel.readbits
* `proc foo(out B: []) { for i in B.domain do B(i) = i; }`
* `proc f(out arg) { if something then arg = 1; }` from SSCA2

**Comment by mppf** [comment]
@vasslitvinov - I think the idea of `proc initViaOut(out arg: ?T)` opting in to the type being passed from the call site is interesting.

**Comment by bradcray** [comment]
I'm also really intrigued by the proposal, though I keep feeling like it seems too good to be true.  I gave myself last night to see whether I could come up with something about it that seemed unworkable or non-orthogonal, but am still liking it this  morning.

**Comment by e-kayrakli** [comment]
Michael has walked me through the general idea proposed here (type-resolving out formals as if they are returned values and not passed arguments). I think it makes sense and improves the behavior of `out` intents. I had only one minor concern about the ambiguity coming from having a generic overload of a function with `out` intent. I would like to have the ambiguity error message to allude to the fact that `out` formal resolution being different than type resolution of formals with other intents. I don't think it is a high priority given the rarity of overloads of functions that only differ by `out` formal types.

**Comment by mppf** [comment]
The design questions are resolved and PR #16990 implements the change so I am closing this issue. We have issue  #17198 for the remaining work of implementing type inference from the call site for cases like `proc f(out arg: ?T)`.
